File: Algorithm/Python/230. Kth Smallest Element in a BST.py
# -*- coding: utf-8 -*-
# @Author: WuLC
# @Date:   2016-08-24 15:23:25
# @Last modified by:   WuLC
# @Last Modified time: 2016-08-24 15:27:20
# @Email: liangchaowu5@gmail.com

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def kthSmallest(self, root, k):
        """
        :type root: TreeNode
        :type k: int
        :rtype: int
        """
        curr, nodes, count = root, [], 0
        while curr or nodes:
            if curr:
                nodes.append(curr)
                curr = curr.left
            else:
                curr = nodes.pop()
                count += 1
                if count == k:
                    return curr.val
                curr = curr.right

File: Algorithm/Java/230. Kth Smallest Element in a BST.java
/*
* @Author: WuLC
* @Date:   2017-03-04 21:24:07
* @Last Modified by:   WuLC
* @Last Modified time: 2017-03-04 21:24:47
* @Email: liangchaowu5@gmail.com
*/


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */



// find recursively
public class Solution 
{
    public int kthSmallest(TreeNode root, int k) 
    {
        int count = countNodes(root.left);
        if (count == k-1) return root.val;
        else if(count > k-1) return kthSmallest(root.left, k);
        else return kthSmallest(root.right, k-count-1);
    }
    
    public int countNodes(TreeNode root)
    {
        if (root == null) return 0;
        return 1+countNodes(root.left)+countNodes(root.right);
    }
}